### PR TITLE
amrfinder & gambit updates

### PR DESCRIPTION
**Pull request by kapsakcj**
~~Draft for now...~~

~~I'm building a validation dataset to test this as well, will post back when it's ready to share~~

Validation dataset is ready to share, the FASTQs & Terra metadata sheet are on our public google bucket. These can be used for validating amrfinderplus outputs in the future, to verify that organism-specific point mutations are detected & "blacklisted genes" are removed for those species.

## :hammer_and_wrench: Changes Being Made

- CI/GHActions changes
  - updated hashes where appropriate
  - added lines to delete unnecessary stuff to allow for more disk space, preventing runners from running out of space when running cromwell & miniwdl

- `tasks/gene_typing/task_amrfinderplus.wdl` 
  - added support for all organisms that amrfinder now supports (they have added a few in the last couple of months)
  - exposed `memory` as optional input to task, default is `16`
- altered `tasks/taxon_id/task_gambit.wdl` to allow for additional `merlin_tag` String outputs. REMINDER the `merlin_tag` string is used for triggering organism-specific tasks and is passed along to `amrfinderplus_nuc` task
  - added K. pneumoniae, K. oxytoca, K. aerogenes, S. pseudintermedius, Streptococcus pyogenes, V. cholerae, Burkholderia cepacia, Burkholderia pseudomallei, C. coli, C. jejuni, Citrobacter freundii, Clostridioides difficile, Enterobacter cloacae, Enterococcus faecalis, Enterococcus hirae, Enterococcus faecium, Serratia marcescens
- All TheiaProk workflows (`workflows/theiaprok/wf_theiaprok_fasta.wdl`, `workflows/theiaprok/wf_theiaprok_illumina_pe.wdl`, `workflows/theiaprok/wf_theiaprok_illumina_se.wdl`, `workflows/theiaprok/wf_theiaprok_ont.wdl`
  - added optional String input `amrfinder_organism` to allow the user to override the `gambit_predicted_taxon` as input for amrfinderplus task. This is useful when GAMBIT does not correctly call the species and you want the organism-specific outputs from amrfinderplus. Useful for enterobacter_cloacae, Clostridioides_difficile, and maybe other species.
- `workflows/utilities/wf_merlin_magic.wdl`
  - split up conditional for running Kleborate into 4 species (pneumoniae, variicola, aerogenes, oxytoca) since `merlin_tag` outputs from GAMBIT task are now broken up into species instead of just the genus `Klebsiella`

## :brain: Context and Rationale

This will allow for additional detection of point mutations, filtering of "blacklisted" genes for the given taxa. Basically everything here: https://github.com/ncbi/amr/wiki/Running-AMRFinderPlus#--organism-option

When I initially wrote the amrfinder task, I did not include the full list of organisms, only the most important/frequently analyzed ones (Salmonella, E. coli, Campy, etc.)

### Locally

Ran TheiaProk_Illumina_PE via miniwdl on one klebsiella pneumoniae, which ran successfully:

```
$ miniwdl run -v ~/github/public_health_bioinformatics/workflows/theiaprok/wf_theiaprok_illumina_pe.wdl samplename=SRR22276936 read1_raw= SRR22276936_1.fastq.gz read2_raw= SRR22276936_2.fastq.gz
```

### Terra

Tested TheiaProk_Illumina_PE workflow on 55 samples in Terra, call caching off, using validation dataset that is comprised of all of the species supported by AMRFinderPlus's `--organism` flag. Terra workflow here: https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/b315e08b-b2b8-47b8-9e1d-d843a6bb71b6

The only hiccups were that GAMBIT could not call the correct species for Enterobacter cloacae & Clostroides difficile. For these 2 species, I ran separate workflows and supplied the correct optional input `amrfinder_organism` to merlin_magic workflow to override the `gambit_predicted_taxon` being used as input

TODOs

- [x] test theiaprok illumina SE workflow with same dataset
Successful workflow here: https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/22740153-ed4f-474c-9ebb-48f0c1359adf
- [x] test at least one sample via TheiaProk_ONT workflow, perhaps a Klebsiella pneumoniae sample
workflow running now: https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/94ea89bf-7142-4e62-bee0-79c36aa9e87c
- [x] test TheiaProk_FASTA workflow with assemblies output from TheiaProk_Illumina_PE workflow
Successful workflow here: https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/1bef0492-45d6-4981-b309-504e10b5b501

## :microscope: Quality checks

Pull Request (PR) checklist:
- [x] Include a description of what is in this pull request in this message.
- [x] The workflow/task has been tested locally and on Terra
- [x] The CI/CD has been adjusted and tests are passing
- [x] Everything follows the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-bb456f34322d4f4db699d4029050481c)

### CODE REVIEW

**Comment by kapsakcj** [comment]
Need to double check on the imact of changing `merlin_tag` for Klebsiella species. Looks like Kleborate was not run on the K. pneumoniae test samples which is unexpected.

- [ ] figure out which Klebsiella species are appropriate to run through kleborate
- [ ] adjust merlin_magic workflow to accept different species(?) or just the Klebsiella pneumoniae species
- [ ] re-test theiaprok wf on K. pneumoniae samples

**Comment by kapsakcj** [comment]
Final tests ran successfully after implementing change to `merlin_tag` string output for Shigella sonnei: https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/bed64fa6-7122-4790-b069-2f99559955e0

**Comment by michellescribner** [comment]
Just a thought: adding the amrfinder_organism input seems like a great way of modifying the merlin_tag if GAMBIT predicts the wrong species. For the qc_check module, I created a similar input called "expected_taxon" that can similarly be used to override the GAMBIT assignment. We might consider consolidating the two inputs since I imagine they would be redundant.

**Comment by sage-wright** [comment]
Definitely agree with Shelly about consolidating the two variables. The code looks good! Running tests here:
https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Wright_PHBG_Sandbox/job_history/d5e77768-aa29-4547-8004-1577bb74e46e
https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Wright_PHBG_Sandbox/job_history/6fac8e00-e3fb-4cc8-8f30-fae919f327e2

**Review by sage-wright — CHANGES_REQUESTED** [review]
Looks good. I'll approve upon my tests succeeded, the merge conflict resolution (md5sums), and (optionally) the consolidation of amrfinderplus_organism & expected_taxon to avoid redundancy

File: workflows/utilities/wf_merlin_magic.wdl
Comment (sage-wright) [inline workflows/utilities/wf_merlin_magic.wdl:1]: You'll also need to add a || merlin_tag == "Vibrio cholerae" on line 324

Comment (sage-wright) [reply workflows/utilities/wf_merlin_magic.wdl:1]: I've done this in my new PR #114

**Comment by michellescribner** [comment]
Testing completed successfully for:

- TheiaProk_Illumina_PE_PHB: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Scribner_Sandbox/job_history/8e21b64a-f38c-4ad4-ba3e-12d069a585fc
- TheiaProk_Illumina_PE_PHB with expected_taxon="clostridiodes_difficile" : https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Scribner_Sandbox/job_history/3684d93c-5294-490f-80ab-8c2d30c31413
- TheiaProk_FASTA_PHB: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Scribner_Sandbox/job_history/50010297-c777-469f-96fa-9165fa482251

**Review by sage-wright — APPROVED** [review]

**Review by kevinlibuit — APPROVED** [review]

**Pull request merged**